### PR TITLE
Add support for process.domain and some other process stuff #5703

### PR DESCRIPTION
**Pull request by cjbarth**
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/5703

### CODE REVIEW

File: node/node.d.ts [hunk 1]
```diff
@@ -626,6 +646,12 @@ declare module "cluster" {
         silent?: boolean;
     }
 
+    export interface Address {
```
Comment (vvakame) [inline node/node.d.ts]: what's this?

File: node/node.d.ts [hunk 2]
```diff
@@ -275,9 +292,12 @@ declare module NodeJS {
         umask(mask?: number): number;
         uptime(): number;
         hrtime(time?:number[]): number[];
+        domain: Domain;
 
         // Worker
         send?(message: any, sendHandle?: any): void;
+        disconnect(): void;
+        connected: boolean;
```
Comment (vvakame) [inline node/node.d.ts]: https://nodejs.org/dist/latest-v4.x/docs/api/process.html#process_process_connected

File: node/node.d.ts [hunk 3]
```diff
@@ -275,9 +292,12 @@ declare module NodeJS {
         umask(mask?: number): number;
         uptime(): number;
         hrtime(time?:number[]): number[];
+        domain: Domain;
 
         // Worker
         send?(message: any, sendHandle?: any): void;
+        disconnect(): void;
```
Comment (vvakame) [inline node/node.d.ts]: https://nodejs.org/dist/latest-v4.x/docs/api/process.html#process_process_disconnect

File: node/node.d.ts [hunk 4]
```diff
@@ -1720,15 +1746,18 @@ declare module "crypto" {
         final(): Buffer;
         final(output_encoding: string): string;
         setAutoPadding(auto_padding: boolean): void;
+        getAuthTag(): Buffer;
```
Comment (vvakame) [inline node/node.d.ts]: https://nodejs.org/dist/latest-v4.x/docs/api/crypto.html#crypto_cipher_getauthtag

File: node/node.d.ts [hunk 5]
```diff
@@ -1720,15 +1746,18 @@ declare module "crypto" {
...
     }
     export function createDecipher(algorithm: string, password: any): Decipher;
     export function createDecipheriv(algorithm: string, key: any, iv: any): Decipher;
     export interface Decipher {
         update(data: Buffer): Buffer;
-        update(data: string, input_encoding?: string, output_encoding?: string): string;
+        update(data: string|Buffer, input_encoding?: string, output_encoding?: string): string;
+        update(data: string|Buffer, input_encoding?: string, output_encoding?: string): Buffer;
```
Comment (vvakame) [inline node/node.d.ts]: https://nodejs.org/dist/latest-v4.x/docs/api/crypto.html#crypto_decipher_update_data_input_encoding_output_encoding

File: node/node.d.ts [hunk 6]
```diff
@@ -1720,15 +1746,18 @@ declare module "crypto" {
...
         final(): Buffer;
         final(output_encoding: string): string;
         setAutoPadding(auto_padding: boolean): void;
+        setAuthTag(tag: Buffer): void;
```
Comment (vvakame) [inline node/node.d.ts]: https://nodejs.org/dist/latest-v4.x/docs/api/crypto.html#crypto_decipher_setauthtag_buffer

File: node/node.d.ts [hunk 7]
```diff
@@ -211,6 +211,23 @@ declare module NodeJS {
...
+        bind(cb: (err: Error, data: any) => any): any;
+        intercept(cb: (data: any) => any): any;
+        dispose(): void;
+
+        addListener(event: string, listener: Function): Domain;
+        on(event: string, listener: Function): Domain;
+        once(event: string, listener: Function): Domain;
+        removeListener(event: string, listener: Function): Domain;
+        removeAllListeners(event?: string): Domain;
+    }
```
Comment (vvakame) [inline node/node.d.ts]: please modify this line
https://github.com/cjbarth/DefinitelyTyped/blob/65b984206300e3cf8884ab07e46400a760aa733a/node/node.d.ts#L1983
likes the following pattern.
https://github.com/cjbarth/DefinitelyTyped/blob/65b984206300e3cf8884ab07e46400a760aa733a/node/node.d.ts#L458

File: node/node.d.ts (hunk 1 above)
Comment (cjbarth) [inline node/node.d.ts]: https://nodejs.org/dist/latest-v4.x/docs/api/cluster.html#cluster_event_listening_1

File: node/node.d.ts (hunk 7 above)
Comment (cjbarth) [inline node/node.d.ts]: Done. See line 1983 below.

File: node/node.d.ts (hunk 1 above)
Comment (vvakame) [inline node/node.d.ts]: :+1:

**Comment by vvakame** [comment]
@cjbarth I think this definition has a problem.
```
        update(data: string|Buffer, input_encoding?: string, output_encoding?: string): string;
        update(data: string|Buffer, input_encoding?: string, output_encoding?: string): Buffer;
```

http://www.typescriptlang.org/Playground#src=class%20Buffer%20%7B%20%7D%0A%0Ainterface%20Temp%20%7B%0A%09update(data%3A%20string%20%7C%20Buffer%2C%20input_encoding%3F%3A%20string%2C%20output_encoding%3F%3A%20string)%3A%20string%3B%0A%09update(data%3A%20string%20%7C%20Buffer%2C%20input_encoding%3F%3A%20string%2C%20output_encoding%3F%3A%20string)%3A%20Buffer%3B%0A%7D%0A%0Avar%20temp%3A%20Temp%3B%0A%0A%2F%2F%20valueA%20is%20'string'%20type%0Alet%20valueA%20%3D%20temp.update(%22%22)%3B%0A%2F%2F%20valueB%20is%20'string'%20type%20too!%0Alet%20valueB%20%3D%20temp.update(new%20Buffer())%3B%0A
please check this sample. If you can fix this issue, please send a new PR ;)

**Comment by cjbarth** [comment]
What do you think of this solution that now seems to work really well with Typescript 1.8?

http://www.typescriptlang.org/Playground#src=class%20Buffer%20%7B%20%7D%0A%0Ainterface%20Temp%20%7B%0A%09update(data%3A%20string%20%7C%20Buffer%2C%20input_encoding%3F%3A%20string%2C%20output_encoding%3F%3A%20string)%3A%20string%20%26%20Buffer%3B%0A%7D%0A%0Avar%20temp%3A%20Temp%3B%0A%0A%2F%2F%20valueA%20is%20'string%20%26%20Buffer'%20type%0Alet%20valueA%20%3D%20temp.update(%22%22)%3B%0A%2F%2F%20valueB%20is%20'string%20%26%20Buffer'%20type%20too!%0Alet%20valueB%20%3D%20temp.update(new%20Buffer())%3B%0A%2F%2F%20valueC%20is%20'string'%20type%0Alet%20valueC%3A%20string%20%3D%20temp.update(%22%22)%3B%0A%2F%2F%20valueD%20is%20'Buffer'%20type%0Alet%20valueD%3A%20Buffer%20%3D%20temp.update(new%20Buffer())%3B%0A

**Comment by vvakame** [comment]
we should use union types. like this `update(data: string | Buffer, input_encoding?: string, output_encoding?: string): string | Buffer;`
or use specialized signature https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#3924-specialized-signatures like this http://goo.gl/Bn0Wj8 (I don't this definition is right in crypto module)

**Pull request merged**